crates/bevy_hierarchy/src/one_to_many/one.rs

```diff
@@ -0,0 +1,200 @@
+#[cfg(feature = "reflect")]
+use bevy_ecs::reflect::{
+    ReflectComponent, ReflectFromWorld, ReflectMapEntities, ReflectVisitEntities,
+    ReflectVisitEntitiesMut,
+};
+use bevy_ecs::{
+    component::{Component, ComponentId},
+    entity::{Entity, VisitEntities, VisitEntitiesMut},
+    event::Events,
+    traversal::Traversal,
+    world::{DeferredWorld, FromWorld, World},
+};
+use core::{fmt::Debug, marker::PhantomData, ops::Deref};
+
+use super::{OneToMany, OneToManyEvent};
+
+/// Represents one half of a one-to-many relationship between an [`Entity`] and some number of other [entities](Entity).
+///
+/// The type of relationship is denoted by the parameter `R`.
+#[derive(Component)]
+#[component(
+    on_insert = Self::associate,
+    on_replace = Self::disassociate,
+)]
+#[cfg_attr(feature = "reflect", derive(bevy_reflect::Reflect))]
+#[cfg_attr(
+    feature = "reflect",
+    reflect(
+        Component,
+        MapEntities,
+        VisitEntities,
+        VisitEntitiesMut,
+        PartialEq,
+        Debug,
+        FromWorld
+    )
+)]
+pub struct ManyToOne<R> {
+    entity: Entity,
+    #[cfg_attr(feature = "reflect", reflect(ignore))]
+    _phantom: PhantomData<fn(&R)>,
+}
+
+impl<R: 'static> ManyToOne<R> {
+    fn associate(mut world: DeferredWorld<'_>, a_id: Entity, _component: ComponentId) {
+        world.commands().queue(move |world: &mut World| {
+            let b = world
+                .get_entity(a_id)
+                .ok()
+                .and_then(|a| a.get::<Self>())
+                .map(|a_relationship| a_relationship.entity)
+                .and_then(|b_id| world.get_entity_mut(b_id).ok());
+
+            let Some(mut b) = b else { return };
+
+            let b_id = b.id();
+
+            let b_points_to_a = b
+                .get::<OneToMany<R>>()
+                .is_some_and(|b_relationship| b_relationship.contains(&a_id));
```

Comment:
For large relationships an `EntityHashSet` would be more appropriate, but `SmallVec` allows us to avoid allocation for small relationship graphs, which I believe is more common (or at least, it's the status-quo). Ideally, we would have some `SmallEntityHashSet` type which used a `SmallVec` optimisation for a small quantity of entities, and then switched to a `EntityHashSet` for larger quantities.

However, I think this would complicate an already challenging PR for an unknown performance win, so I'd rather save this for a future PR.

Reply:
I think the current implementation is fine and good. But in case you do want to explore this further, please see https://docs.rs/tinyset/latest/tinyset/ . Basically, we'd want a `tinyset::Set64` that is optimized for `Entity` types in a way where small entity ids with small generation numbers would end up being small numbers, possibly meaning that we'd need to do Morton encoding on the entity id. But that's a very specific optimization so should be a separate pull with measurable speed benefits.